lib/semian/grpc.rb

```diff
@@ -0,0 +1,80 @@
+require 'semian/adapter'
+require 'grpc'
+
+module GRPC
+  GRPC::BadStatus.include(::Semian::AdapterError)
+
+  class SemianError < GRPC::BadStatus
+    def initialize(semian_identifier, *args)
+      super(*args)
+      @semian_identifier = semian_identifier
+    end
+  end
+
+  ResourceBusyError = Class.new(SemianError)
+  CircuitOpenError = Class.new(SemianError)
+end
+
+module Semian
+  module GRPC
+    include Semian::Adapter
+
+    ResourceBusyError = ::GRPC::ResourceBusyError
+    CircuitOpenError = ::GRPC::CircuitOpenError
+
+    def initialize(*args)
+      @host = args.first
+      set_raw_semian_options(args.last)
+      args.pop
+      super(*args)
+    end
+
+    def semian_identifier
+      @semian_identifier ||= :"grpc_#{@host}"
+    end
+
+    DEFAULT_ERRORS = [
+      ::GRPC::Unavailable,
+      ::GRPC::Core::CallError,
+      ::GRPC::BadStatus,
+    ].freeze
+
+    class << self
+      attr_accessor :exceptions
+
+      def reset_exceptions
+        self.exceptions = Semian::GRPC::DEFAULT_ERRORS.dup
+      end
+    end
+
+    Semian::GRPC.reset_exceptions
+
+    def resource_exceptions
+      Semian::GRPC.exceptions
+    end
+
+    def request_response(*args)
```

Comment:
From a quick read of https://github.com/grpc/grpc/blob/master/src/ruby/ext/grpc/rb_channel.c it appears the connection can be established in the background in another thread https://github.com/grpc/grpc/pull/9986, possibly also via `Core::Channel.new` somehow.

Given calls don't trigger connections, if they only ever happen in another thread, maybe Semian doesn't need to care since slow or failing connect attempts won't influence call latency. There seems to be [a TODO](https://github.com/grpc/grpc/blob/a786b7415aaaf7247293f4116007aa451652be06/src/ruby/ext/grpc/rb_channel.c#L733) to change this though.